dataless-classifier/README.md

```diff
@@ -0,0 +1,59 @@
+# CogComp-DatalessClassifier
+
+Some key points:
+- The Main classes for the Dataless Annotators are :-
+  * **ESADatalessAnnotator** for the ESA-based Dataless Annotator
+  * **W2VDatalessAnnotator** for the Word2Vec-based Dataless Annotator
+- Dataless Annotators add the **ESA-Dataless** and **W2V-Dataless** views to the input TextAnnotation respectively, and it requires the presence of a **TOKENS** view with the end-user's desired Tokenization.
+- Since Labels/Topics are inferred at the Document-Level, all topic annotations span the entire document.
+- Sample invocation has been provided in the main functions of each annotator.
+- Both annotators load up embeddings in memory, and thus can easily consume upto **10GB RAM**.
+
+
+## Label Hierarchy
+Dataless Classification requires the end-user to specifcy a Label hierarchy (with label descriptions), which it classifies into. The Label hierarchy is inputted using a very specific format:
+* **labelNamePath**: Specify your label id to label name mapping here in the `labelID \t labelName` format 
+  (label id can be any ID specific to your system, however we use the label name itself as ID in our sample hierachy for readibility)
+* **labelHierarchyPath**: The first line of this file should contain tab-separated list of Top-Level nodes in the hierarchy (i.e. the ones directly connected to the root). Then, every following line should specify the connections in the hierachy in the `parentLabelID \t childLabelID1 \t childLabelID2 \t ...` format.
+* **labelDescPath**: Dataless' performance hinges on good label descriptions, which you specify in this file in the `labelID \t labelDescription` format.
+
+We provide a sample 20newsgroups hierarchy with label descriptions inside data/hierarchy/20newsgroups, where :-
+* idToLabelNameMap.txt should be used as labelNamePath
+* parentChildIdMap.txt should be used as labelHierarchyPath
+* labelDesc\_Kws\_simple.txt should be used as labelDescPath
+
+We also provide improved 20newsgroups label descriptions in *labelDesc\_Kws\_embellished.txt* which corresponds to the label descriptions used in [2], whereas the *labelDesc\_Kws\_simple.txt* corresponds to the label descriptions used in [1].
+
+## Embeddings
+Dataless Annotators also require the corresponding embeddings, i.e. ESA embeddings for the ESADatalessAnnotator, and Word2Vec embeddings for the W2VDatalessAnnotator. If you are using the project as a dependency, you'll have to add the following dependencies too :-
+
+	<dependency>
+		<groupId>edu.illinois.cs.cogcomp</groupId>
+		<artifactId>esaEmbedding</artifactId>
+		<version>1.0</version>
+	</dependency>
+	<dependency>
+		<groupId>edu.illinois.cs.cogcomp</groupId>
+		<artifactId>w2vEmbedding-100</artifactId>
+		<version>1.0</version>
+	</dependency>
```

Comment:
Ideally we should use [Datastore](https://github.com/CogComp/cogcomp-datastore) for loading large resources. This would make things lazy (i.e. they would get downloaded (and cached) only when we call them). Example usages are in the links I posted about the similarity package.

Reply:
@danyaljj, @mssammon  .. Some quick background:
1. Dataless Classification concerns with Document Embeddings, and thus segmentation of documents and composition of word-level embeddings are important -- For word-level embeddings, you can choose either a Sparse Embedding (like ESA), or a Dense Embedding (like Word2Vec).
2. The similarity package has more or less copied the code from the legacy Dataless package, and thus the segmentation and compositionality related components are hard-coded. Moreover, it doesn't unify the interface for ESA and Word2Vec embeddings, since one of them is a Sparse Embedding, and the other is a Dense Embedding -- which probably makes sense for a Similarity package.
3. On the other hand, I had refactored most of the legacy Dataless Classification codebase to make it more friendly for research. For instance:
     - I unified the interface for ESA and W2V embeddings since the Dataless Classifier code works only on Sparse Embeddings, and designed it in such a way that most of the segmentation and compositionality related components have been refactored out into separate functions -- thus providing flexibility for user-desired segmentation and composition functions.
    - I wrote generic DenseVector and SparseVector implementations, thus allowing embeddings to be associated with arbitrary Document components, for instance Entities, Relations, Types, Paragraphs etc.

So, this is where things get a bit muddy. There definitely is some intersection, both on the code-side and the project-scope side, but there is no clear solution without major refactoring efforts. Some ideas to deal with this, with their cons:

1. Refactor out the Document Embedding computation portion from Dataless-Classifier and incorporate it into the similarity package. Some issues with this approach: 
    - The document embedding computation code in Dataless-Classifier was written with Dataless-Classifier in mind, and is not generic enough. For instance, even if the word-level embedding is a Dense Embedding, we emit Document-level Sparse Embeddings since that's what the Dataless Classifier works on -- This might not be desirable for a generic Similarity Package, for it should emit a Document-level Dense Embedding.
    - Dataless-Classifier's AEmbedding class is analogous to the Embedding class in the Similarity package, with the difference that AEmbedding works with Sparse Embeddings, and Embedding works with double arrays (Dense Embedding) .. so, if Embedding is replaced with AEmbedding in the similarity package, most of its code will have to be adapted accordingly. That adaptation will still be fine if the scope of the Similarity package is just to provide Similarities, and not the underlying embeddings themselves.

2. Revert back the Dataless-Classifier to its legacy code, and use just similarities from the similarity package. Some issues with this approach:
    - We end up losing all the work that has gone into making the Dataless package easier for expansion and future research.

3. Let them coexist side-by-side for now.

Reply:
Hi @shatu . 

Thanks for clarifying the issues. 

For this PR, I think we should move on with the transition, without much changes in the similarity package (item 3 above). 
I don't like (2).  I like most of 3, but I have to study it too so that I completely understand what you're saying and definitely not for this PR. 
So my suggestions is let's move on with (3) and we will get back to (1). 

What do you think?

Reply:
@danyaljj: Sounds good! I'll just address the 2 leftover bullet-points from your previous code-review then!

Reply:
@danyaljj .. Apart from the DataStore issue that I mentioned (that it only uploads upto 2GB sized file), I'm done from my end.